cuda/dot_kernels.cuh

```diff
@@ -0,0 +1,224 @@
+#pragma once
+
+#include <cinttypes>
+
+
+#include <cooperative_groups.h>
+#include <cublas_v2.h>
+#include <accessor/range.hpp>
+#include <accessor/reduced_row_major.hpp>
+
+
+#include "atomics.cuh"
+#include "kernel_utils.cuh"
+#include "utils.cuh"
+
+
+constexpr int grids_per_sm{32};
+constexpr int dot_block_size{1024};
+
+
+class myBlasHandle {
+public:
+    myBlasHandle()
+    {
+        CUDA_CALL(cudaGetDeviceProperties(&device_prop_, 0));
+        CUDA_CALL(cudaMalloc(&device_storage_, device_storage_size_bytes_));
+    }
+
+    const cudaDeviceProp &get_device_property() const { return device_prop_; }
+
+    template <typename T>
+    T *get_device_value_ptr()
+    {
+        static_assert(sizeof(T) < device_storage_size_bytes_,
+                      "The expected type is too large for the device storage!");
+        return reinterpret_cast<T *>(device_storage_);
+    }
+
+    ~myBlasHandle() { cudaFree(&device_storage_); }
+
+private:
+    static constexpr std::size_t device_storage_size_bytes_{16};
+    cudaDeviceProp device_prop_;
+    void *device_storage_;
+};
+
+
+namespace kernel {
+
+
+namespace cg = cooperative_groups;
+
+template <typename ValueType>
+__global__ __launch_bounds__(1) void init_res(ValueType *__restrict__ res)
+{
+    *res = ValueType{0};
+}
+
+template <std::int64_t block_size, typename ValueType>
+__global__ __launch_bounds__(block_size) void dot(
+    const std::int32_t n, const ValueType *__restrict__ x,
+    const std::int32_t x_stride, const ValueType *__restrict__ y,
+    const std::int32_t y_stride, ValueType *__restrict__ res)
+{
+    // Here, using int32 is fine since input & stride are also in int32
+    using index_type = std::int32_t;
+
+    const index_type start = blockIdx.x * blockDim.x + threadIdx.x;
+    const index_type increment = blockDim.x * gridDim.x;
+
+    ValueType local_result{};
+    const auto group = cg::this_thread_block();
+    const auto local_id = group.thread_rank();
+
+    __shared__ char shared_impl[sizeof(ValueType) * block_size];
+    auto shared = reinterpret_cast<ValueType *>(shared_impl);
+
+    for (index_type idx = start; idx < n; idx += increment) {
+        const auto x_val = x[idx * x_stride];
+        const auto y_val = y[idx * y_stride];
+        local_result += x_val * y_val;
+    }
+    shared[local_id] = local_result;
+    reduce(group, shared, [](ValueType a, ValueType b) { return a + b; });
+    if (local_id == 0) {
+        atomic_add(res, shared[local_id]);
+    }
+}
+
+// Currently, x and y need to be 2D to also have a proper stride parameter
+// (lowest-dim must always be 0)
+template <std::int64_t block_size, typename XRange, typename YRange,
+          typename ResType>
+__global__ __launch_bounds__(block_size) void acc_dot(XRange x, YRange y,
```

Comment:
I think if this were to also be an example, adding some documentation and detailing the few differences with the base implementation could be useful.